### PR TITLE
REF: implement maybe_promote_scalar

### PR DESCRIPTION
**Pull request by jbrockmendel**
Narrowing down the set of cases where we have to worry about a non-scalar fill_value.  I'm pretty sure we can narrow it down further, but saving that for a separate step as it has some complications for EA corner cases.

cc @jreback

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -353,6 +353,12 @@ def maybe_promote(dtype, fill_value=np.nan):
             fill_value = np.nan
             dtype = np.dtype(np.object_)
 
+    return maybe_promote_scalar(dtype, fill_value)
+
+
+def maybe_promote_scalar(dtype, fill_value=np.nan):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: can you add a doc-string / types

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: will do

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: added docstring, annotations i need to coordinate with others about implementing correct types in pd._typing

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jbrockmendel** [comment]
ping

**Comment by jbrockmendel** [comment]
gentle ping.  we're getting close to being done with the maybe_promote saga

**Comment by jbrockmendel** [comment]
Unnecessary following #29416, closing.

**Pull request closed without merging**